design/multiple-backup-selectors_design.md

```diff
@@ -0,0 +1,70 @@
+# Ensure support for backing up resources based on multiple labels
+## Abstract
+As of today Velero supports filtering of resources based on single label selector per backup. It is desired that Velero
+support backing up of resources based on multiple labels (OR logic).
+
+**Note:** This solution is required because kubernetes label selectors only allow AND logic of labels.
+
+## Background
+Currently, Velero's Backup API has a spec field `LabelSelector` which helps in filtering of resources based on
+a **single** label value per backup request. For instance, if the user specifies the `Backup.Spec.LabelSelector` as
+`data-protection-app: true`, Velero will grab all the resources that possess this label and perform the backup
+operation on them. The `LabelSelector` field does not accept more than one labels, and thus if the user want to take
+backup for resources consisting of a label from a set of labels (label1 OR label2 OR label3) then the user needs to
+create multiple backups per label rule. It would be really useful if Velero Backup API could respect a set of
+labels (OR Rule) for a single backup request.
+
+Related Issue: https://github.com/vmware-tanzu/velero/issues/1508
+
+## Goals
+- Enable support for backing up resources based on multiple labels (OR Logic) in a single backup config.
+
+## Use Case/Scenario
+Let's say as a Velero user you want to take a backup of secrets, but all these secrets do not have one single consistent
+label on them. We want to take backup of secrets having any one label in `app=gdpr`, `app=wpa` and `app=ccpa`. Here
+we would have to create 3 instances of backup for each label rule. This can become cumbersome at scale.
+
+## High-Level Design
+### Addition of `LabelSelectors`(_plural_) spec to Velero Backup API
+For Velero to back up resources if they consist of any one label from a set of labels, we would like to add a new spec
+field `LabelSelectors` which would enable user to specify them. The Velero backup command would somewhat look like:
+
+`Velero create backup multi-label-example --selectors 'app=gdpr', 'app=wpa', 'app=ccpa'`
+
+In the above command (`velero --help` will display),
+`--selectors is LabelSelectors  Only backup resources matching any one label from the set of label selectors`
+
+**Note:** This approach will **not** be changing any current behavior related to Backup API spec `LabelSelector`. Rather we
+propose that the label in `LabelSelector` spec and labels in `LabelSelectors` should be concatenated  and then used as a
+criteria(resources matching any one label) to grab the resources to be backed up. For instance, if the backup command
+looks like:
```

Comment:
Hmm. I'm not sure that concatenating is the right approach here, since that changes the meaning of the current (singular) label selector field. I'm wondering whether it might be cleaner (and easier to implement) to just treat them as separate fields. If the singular field is present, then a match requires that label. If the plural field is present, then a match requires one of the entries to match. If both are present, then it's effectively an *and* between them (but presumably most real-world cases won't use both).

Reply:
@kaovilai "introduce another field" -- meaning in addition to the proposed new `LabelSelectors`? i.e. we have `LabelSelector`, `OrLabelSelectors` and `AndLabelSelectors`? If there's a need for that, it could be considered as well -- but this makes my suggestion of "Treat the fields as separate, don't concatenate them" even more important. Is there a need for the ability of including *and* behavior for 3 or more label selectors?

Reply:
Possibly.

Here's one
Backup: east database backup
app=database
region=east

Reply:
Yeah that's a good point. I'm in favor of what Scott is suggesting here. Thoughts @shubham-pampattiwar ?

Reply:
@dymurray yeah I agree, +1 to what @sseago is suggesting. I will update the design proposal with keeping the `labelSelector` and `labelSelectors` as separate fields.